### PR TITLE
Login logout auth

### LINKED ISSUE #3: login, logout, create user and delete user

### LINKED ISSUE #7: authorize for unspecified user, user, leader, admin

### PR DESCRIPTION
**Pull request by quocbang**

### CODE REVIEW

File: server/internal/repositories/orm/models/user.go [hunk 1]
```diff
@@ -1,13 +1,117 @@
...
+		return nil, err
+	}
+
+	claims, ok := JWTToken.Claims.(*JwtCustomClaims)
+	if !ok || !JWTToken.Valid {
+		return nil, repoErrors.Error{
+			Code:    repoErrors.Code_INVALID_TOKEN,
+			Details: "token invalid",
+		}
+	}
```
Comment (leducthai) [inline server/internal/repositories/orm/models/user.go:114]: this is redundant. these verifications are done inside the `jwt.ParseWithClaims` operation you can add the custom error as follow 
```
if err != nil {
		if e, ok := err.(*jwt.ValidationError); ok {
			if e.Errors == jwt.ValidationErrorExpired {
				return nil, repoErrors.Error{
					Code:    repoErrors.Code_TOKEN_EXPIRED,
					Details: err.Error(),
				}
			}
                        return nil, repoErrors.Error{
			         Code:    repoErrors.Code_INVALID_TOKEN,
			         Details: "token invalid",
		        }
		}
		return nil, err
	}
```
and add a variable of `claims := JwtCustomClaims{}` then pass the pointer in `jwt.ParseWithClaims`

File: server/internal/repositories/servicestest/internal/suite/build.go [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+type SuiteConfig struct {
+	*suite.Suite
+	DB   *gorm.DB
+	Repo repositories.Repositories
+
+	relativeModels       []models.Models
+	clearDataForEachTest bool
+}
+
+type NewSuiteParameters struct {
```
Comment (leducthai) [inline server/internal/repositories/servicestest/internal/suite/build.go:25]: If put "New" in front of the name it sounds like it will make a new object

**Review by leducthai — COMMENTED** [review]
@quocbang please resolve the threads

File: server/internal/repositories/servicestest/internal/suite/build.go [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+		relativeModels:       params.RelativeModels,
+		Suite:                &suite.Suite{},
+		clearDataForEachTest: params.ClearDataForEachTest,
+	}
+}
+
+var logger *zap.Logger
+
+func (s *SuiteConfig) SetupSuite() {
+	ranDomSeed := fake.UUID()
```
Comment (leducthai) [inline server/internal/repositories/servicestest/internal/suite/build.go:43]: I think you didn't use the `ranDomSeed`

Comment (quocbang) [reply server/internal/repositories/servicestest/internal/suite/build.go:43]: something wrong?

File: server/internal/repositories/servicestest/internal/suite/build.go [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+}
+
+func (s *SuiteConfig) ClearData() error {
+	for _, m := range s.relativeModels {
+		if err := s.DB.Where("1=1").Delete(m).Error; err != nil {
+			return err
+		}
+	}
+	return nil
+}
```
Comment (leducthai) [inline server/internal/repositories/servicestest/internal/suite/build.go:70]: also clear the Redis server using `s.rd.FlushDB(context.Background())' (note: add `*redis.Client` in `SuiteConfig`

File: server/internal/repositories/servicestest/internal/suite/build.go (hunk 3 above)
Comment (leducthai) [reply server/internal/repositories/servicestest/internal/suite/build.go:43]: if you want to use the seed you must call `fake.Seed(randomSeed)` and random seed should be int64
btw the name should be random seed

Comment (quocbang) [reply server/internal/repositories/servicestest/internal/suite/build.go:43]: are you sure? `gofakeit.Seed` didn't return anything, and we need a random number or string to make sure each test is each different

image

**Comment by quocbang** [comment]
@leducthai don't forget to remove the ready-review tag after each review

**Review by leducthai — APPROVED** [review]

**Comment by leducthai** [comment]
@quocbang let's merge this MR for now I need some content in here

**Pull request merged**